### ISSUE TITLE
Overriding a .scm file?

### ISSUE THREAD

**Issue by keithbowes** [uim/uim#172]
The elatin IM is good, but it has some weird conflicts, like ~s, which can be the Turkish s with a tail or §.  For some reason, uim always gives the second (without giving a choice of which to use like e.g. ipa-x-sampa does).  I've tried putting a modified version of elatin-rules in ~/.local/share/uim, but it still uses the one in /usr/share/uim.  Do I have to move my modified version to /usr/share/uim (which isn't a safe thing to do)?

The customs directory also doesn't work, because it seems to just be used for a limited number of settings, not a way to modify the rules.

**Comment by kou** [comment]
Could you try `~/.uim`?

**Comment by keithbowes** [comment]
It didn't work, but thanks anyway.  I'm copied them from ~/.uim to /usr/share/uim and now it works fine.  I wish there were a way to do it without root access.

**Comment by kou** [comment]
1. Did you define `LIBUIM_VANILLA` environment variable? https://github.com/uim/uim/blob/master/scm/init.scm#L139
2. Did you change your uid/gid? https://github.com/uim/uim/blob/master/scm/init.scm#L78
3. Did you define `LIBUIM_USER_SCM_FILE` environment variable? https://github.com/uim/uim/blob/master/scm/init.scm#L82

**Comment by keithbowes** [comment]
> 1. Did you define `LIBUIM_VANILLA` environment variable? https://github.com/uim/uim/blob/master/scm/init.scm#L139

No.  When I tried, only the "direct" input method was still there.

> 
>     2. Did you change your uid/gid? https://github.com/uim/uim/blob/master/scm/init.scm#L78

No.  I'd have to look into even how to do that.  I'm guessing that's what /sbin/usermod and /sbin/groupmod are for, but I haven't tried them.

> 
>     3. Did you define `LIBUIM_USER_SCM_FILE` environment variable? https://github.com/uim/uim/blob/master/scm/init.scm#L82

Not at the moment.  I used to, to try to get an XDG-compatible directory structure, but it didn't work (it seems to create ~/.uim.d regardless).  I tried again (reinstalling uim so that my modifications to /usr/share/uim would be overridden) and my modifications to the .scm files still weren't recognized.

**Comment by yamam** [comment]
You do not necessarily need to replace the file under `/usr/share/uim`.

`~/.uim` is not used as a search path for overriding `elatin-rules.scm`, but it is loaded after the input method modules. Therefore, it can load `elatin.scm` and replace the rule set used by ELatin.

For example, the following changes `~s` to produce `ş`:

```scheme
;; ~/.uim

(require "elatin.scm")

(define replace-elatin-rule
  (lambda (rules key output)
    (let ((key-seq
           (list (map string (string->list key)))))
      (cons (list key-seq (list output))
            (filter (lambda (rule)
                      (not (equal? (car rule) key-seq)))
                    rules)))))

(define my-elatin-rules
  (replace-elatin-rule
   elatin-rules-latin-prefix
   "~s"
   "ş"))

(set! elatin-rules 'my-elatin-rules)
```

After restarting the input context, pressing `~` followed by `s` in ELatin produces `ş`.
